File: backend-app/src/main/java/com/gradientgeeks/ageis/backendapp/service/PayloadEncryptionService.java
package com.gradientgeeks.ageis.backendapp.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import javax.crypto.Cipher;
import javax.crypto.SecretKey;
import javax.crypto.spec.GCMParameterSpec;
import java.nio.charset.StandardCharsets;
import java.util.Base64;

/**
 * Service for encrypting and decrypting request/response payloads using AES-256-GCM.
 */
@Service
public class PayloadEncryptionService {
    
    private static final Logger logger = LoggerFactory.getLogger(PayloadEncryptionService.class);
    private static final String AES_GCM_ALGORITHM = "AES/GCM/NoPadding";
    private static final int GCM_TAG_LENGTH = 128;
    private static final int ENCRYPTED_PAYLOAD_VERSION = 1;
    
    @Autowired
    private SessionKeyService sessionKeyService;
    
    @Autowired
    private ObjectMapper objectMapper;
    
    /**
     * Decrypts an encrypted payload using the session key.
     */
    public String decryptPayload(EncryptedPayload encryptedPayload, String sessionId) throws Exception {
        logger.debug("Decrypting payload for session: {}", sessionId);
        
        // Validate version
        if (encryptedPayload.getVersion() != ENCRYPTED_PAYLOAD_VERSION) {
            throw new IllegalArgumentException("Unsupported encryption version: " + encryptedPayload.getVersion());
        }
        
        // Get session key
        SecretKey sessionKey = sessionKeyService.getSessionKey(sessionId);
        if (sessionKey == null) {
            throw new IllegalStateException("Session key not found for session: " + sessionId);
        }
        
        // Decode IV and ciphertext
        byte[] iv = Base64.getDecoder().decode(encryptedPayload.getIv());
        byte[] ciphertext = Base64.getDecoder().decode(encryptedPayload.getCiphertext());
        
        // Initialize cipher
        Cipher cipher = Cipher.getInstance(AES_GCM_ALGORITHM);
        GCMParameterSpec gcmSpec = new GCMParameterSpec(GCM_TAG_LENGTH, iv);
        cipher.init(Cipher.DECRYPT_MODE, sessionKey, gcmSpec);
        
        // Add associated data if provided
        if (encryptedPayload.getAad() != null) {
            cipher.updateAAD(encryptedPayload.getAad().getBytes(StandardCharsets.UTF_8));
        }
        
        // Decrypt the payload
        byte[] decryptedBytes = cipher.doFinal(ciphertext);
        String decryptedPayload = new String(decryptedBytes, StandardCharsets.UTF_8);
        
        logger.debug("Successfully decrypted payload ({} bytes)", decryptedBytes.length);
        return decryptedPayload;
    }
    
    /**
     * Encrypts a response payload using the session key.
     */
    public EncryptedPayload encryptPayload(String payload, String sessionId, String aad) throws Exception {
        logger.debug("Encrypting payload for session: {}", sessionId);
        
        // Get session key
        SecretKey sessionKey = sessionKeyService.getSessionKey(sessionId);
        if (sessionKey == null) {
            throw new IllegalStateException("Session key not found for session: " + sessionId);
        }
        
        // Generate random IV
        byte[] iv = new byte[12]; // 96 bits for GCM
        new java.security.SecureRandom().nextBytes(iv);
        
        // Initialize cipher
        Cipher cipher = Cipher.getInstance(AES_GCM_ALGORITHM);
        GCMParameterSpec gcmSpec = new GCMParameterSpec(GCM_TAG_LENGTH, iv);
        cipher.init(Cipher.ENCRYPT_MODE, sessionKey, gcmSpec);
        
        // Add associated data if provided
        if (aad != null) {
            cipher.updateAAD(aad.getBytes(StandardCharsets.UTF_8));
        }
        
        // Encrypt the payload
        byte[] payloadBytes = payload.getBytes(StandardCharsets.UTF_8);
        byte[] ciphertext = cipher.doFinal(payloadBytes);
        
        // Create encrypted payload object
        EncryptedPayload encryptedPayload = new EncryptedPayload();
        encryptedPayload.setVersion(ENCRYPTED_PAYLOAD_VERSION);
        encryptedPayload.setIv(Base64.getEncoder().encodeToString(iv));
        encryptedPayload.setCiphertext(Base64.getEncoder().encodeToString(ciphertext));
        encryptedPayload.setAlgorithm("AES-256-GCM");
        encryptedPayload.setAad(aad);
        
        logger.debug("Successfully encrypted payload ({} bytes)", payloadBytes.length);
        return encryptedPayload;
    }
    
    /**
     * Decrypts a secure request and returns the original request object.
     */
    public <T> T decryptSecureRequest(SecureRequest secureRequest, String sessionId, Class<T> requestClass) throws Exception {
        // Validate metadata
        RequestMetadata metadata = secureRequest.getMetadata();
        if (!sessionId.equals(metadata.getSessionId())) {
            throw new IllegalArgumentException("Session ID mismatch");
        }
        
        // Decrypt payload
        String decryptedJson = decryptPayload(secureRequest.getEncryptedPayload(), sessionId);
        
        // Parse JSON to object
        return objectMapper.readValue(decryptedJson, requestClass);
    }
    
    /**
     * Creates an encrypted response.
     */
    public SecureResponse createSecureResponse(Object responseObject, String sessionId, int statusCode) throws Exception {
        // Convert response to JSON
        String responseJson = objectMapper.writeValueAsString(responseObject);
        
        // Create response metadata
        ResponseMetadata metadata = new ResponseMetadata();
        metadata.setStatusCode(statusCode);
        metadata.setTimestamp(System.currentTimeMillis());
        metadata.setSessionId(sessionId);
        
        // Create AAD from metadata
        String aad = statusCode + "|" + metadata.getTimestamp() + "|" + sessionId;
        
        // Encrypt payload
        EncryptedPayload encryptedPayload = encryptPayload(responseJson, sessionId, aad);
        
        // Create secure response
        SecureResponse secureResponse = new SecureResponse();
        secureResponse.setEncryptedPayload(encryptedPayload);
        secureResponse.setMetadata(metadata);
        
        return secureResponse;
    }
    
    // Data classes
    public static class EncryptedPayload {
        private int version;
        private String iv;
        private String ciphertext;
        private String algorithm;
        private String aad;
        
        // Getters and setters
        public int getVersion() { return version; }
        public void setVersion(int version) { this.version = version; }
        
        public String getIv() { return iv; }
        public void setIv(String iv) { this.iv = iv; }
        
        public String getCiphertext() { return ciphertext; }
        public void setCiphertext(String ciphertext) { this.ciphertext = ciphertext; }
        
        public String getAlgorithm() { return algorithm; }
        public void setAlgorithm(String algorithm) { this.algorithm = algorithm; }
        
        public String getAad() { return aad; }
        public void setAad(String aad) { this.aad = aad; }
    }
    
    public static class RequestMetadata {
        private String method;
        private String uri;
        private long timestamp;
        private String sessionId;
        
        // Getters and setters
        public String getMethod() { return method; }
        public void setMethod(String method) { this.method = method; }
        
        public String getUri() { return uri; }
        public void setUri(String uri) { this.uri = uri; }
        
        public long getTimestamp() { return timestamp; }
        public void setTimestamp(long timestamp) { this.timestamp = timestamp; }
        
        public String getSessionId() { return sessionId; }
        public void setSessionId(String sessionId) { this.sessionId = sessionId; }
    }
    
    public static class ResponseMetadata {
        private int statusCode;
        private long timestamp;
        private String sessionId;
        
        // Getters and setters
        public int getStatusCode() { return statusCode; }
        public void setStatusCode(int statusCode) { this.statusCode = statusCode; }
        
        public long getTimestamp() { return timestamp; }
        public void setTimestamp(long timestamp) { this.timestamp = timestamp; }
        
        public String getSessionId() { return sessionId; }
        public void setSessionId(String sessionId) { this.sessionId = sessionId; }
    }
    
    public static class SecureRequest {
        private EncryptedPayload encryptedPayload;
        private RequestMetadata metadata;
        
        // Getters and setters
        public EncryptedPayload getEncryptedPayload() { return encryptedPayload; }
        public void setEncryptedPayload(EncryptedPayload encryptedPayload) { this.encryptedPayload = encryptedPayload; }
        
        public RequestMetadata getMetadata() { return metadata; }
        public void setMetadata(RequestMetadata metadata) { this.metadata = metadata; }
    }
    
    public static class SecureResponse {
        private EncryptedPayload encryptedPayload;
        private ResponseMetadata metadata;
        
        // Getters and setters
        public EncryptedPayload getEncryptedPayload() { return encryptedPayload; }
        public void setEncryptedPayload(EncryptedPayload encryptedPayload) { this.encryptedPayload = encryptedPayload; }
        
        public ResponseMetadata getMetadata() { return metadata; }
        public void setMetadata(ResponseMetadata metadata) { this.metadata = metadata; }
    }
}

File: sfe/sfe-client/src/main/java/com/gradientgeeks/aegis/sfe_client/encryption/PayloadEncryptionService.kt
package com.gradientgeeks.aegis.sfe_client.encryption

import android.util.Base64
import android.util.Log
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec

/**
 * Service for encrypting and decrypting request/response payloads using AES-256-GCM.
 * Provides authenticated encryption with associated data (AEAD).
 */
class PayloadEncryptionService {
    companion object {
        private const val TAG = "PayloadEncryptionService"
        private const val AES_GCM_ALGORITHM = "AES/GCM/NoPadding"
        private const val GCM_IV_LENGTH = 12 // 96 bits recommended for GCM
        private const val GCM_TAG_LENGTH = 128 // 128 bits for authentication tag
        private const val ENCRYPTED_PAYLOAD_VERSION = 1
    }
    
    private val secureRandom = SecureRandom()
    
    /**
     * Encrypts a payload using AES-256-GCM with the provided session key.
     * 
     * @param payload The data to encrypt (as string)
     * @param sessionKey The AES-256 session key
     * @param associatedData Optional additional authenticated data (AAD)
     * @return Encrypted payload with metadata, or null on failure
     */
    fun encryptPayload(
        payload: String,
        sessionKey: SecretKey,
        associatedData: String? = null
    ): EncryptedPayload? {
        return try {
            // Generate random IV
            val iv = ByteArray(GCM_IV_LENGTH)
            secureRandom.nextBytes(iv)
            
            // Initialize cipher
            val cipher = Cipher.getInstance(AES_GCM_ALGORITHM)
            val gcmSpec = GCMParameterSpec(GCM_TAG_LENGTH, iv)
            cipher.init(Cipher.ENCRYPT_MODE, sessionKey, gcmSpec)
            
            // Add associated data if provided
            associatedData?.let {
                cipher.updateAAD(it.toByteArray(StandardCharsets.UTF_8))
            }
            
            // Encrypt the payload
            val payloadBytes = payload.toByteArray(StandardCharsets.UTF_8)
            val ciphertext = cipher.doFinal(payloadBytes)
            
            // Create encrypted payload object
            val encryptedPayload = EncryptedPayload(
                version = ENCRYPTED_PAYLOAD_VERSION,
                iv = Base64.encodeToString(iv, Base64.NO_WRAP),
                ciphertext = Base64.encodeToString(ciphertext, Base64.NO_WRAP),
                algorithm = "AES-256-GCM",
                aad = associatedData
            )
            
            Log.d(TAG, "Successfully encrypted payload (${payloadBytes.size} bytes)")
            encryptedPayload
            
        } catch (e: Exception) {
            Log.e(TAG, "Failed to encrypt payload", e)
            null
        }
    }
    
    /**
     * Decrypts a payload using AES-256-GCM with the provided session key.
     * 
     * @param encryptedPayload The encrypted payload object
     * @param sessionKey The AES-256 session key
     * @return Decrypted payload as string, or null on failure
     */
    fun decryptPayload(
        encryptedPayload: EncryptedPayload,
        sessionKey: SecretKey
    ): String? {
        return try {
            // Validate version
            if (encryptedPayload.version != ENCRYPTED_PAYLOAD_VERSION) {
                Log.e(TAG, "Unsupported encryption version: ${encryptedPayload.version}")
                return null
            }
            
            // Decode IV and ciphertext
            val iv = Base64.decode(encryptedPayload.iv, Base64.NO_WRAP)
            val ciphertext = Base64.decode(encryptedPayload.ciphertext, Base64.NO_WRAP)
            
            // Initialize cipher
            val cipher = Cipher.getInstance(AES_GCM_ALGORITHM)
            val gcmSpec = GCMParameterSpec(GCM_TAG_LENGTH, iv)
            cipher.init(Cipher.DECRYPT_MODE, sessionKey, gcmSpec)
            
            // Add associated data if provided
            encryptedPayload.aad?.let {
                cipher.updateAAD(it.toByteArray(StandardCharsets.UTF_8))
            }
            
            // Decrypt the payload
            val decryptedBytes = cipher.doFinal(ciphertext)
            val decryptedPayload = String(decryptedBytes, StandardCharsets.UTF_8)
            
            Log.d(TAG, "Successfully decrypted payload (${decryptedBytes.size} bytes)")
            decryptedPayload
            
        } catch (e: Exception) {
            Log.e(TAG, "Failed to decrypt payload", e)
            null
        }
    }
    
    /**
     * Creates a secure request envelope containing both encrypted and metadata.
     * 
     * @param payload Original request payload
     * @param sessionKey Session key for encryption
     * @param requestMetadata Additional metadata (method, URI, etc.)
     * @return Secure request envelope
     */
    fun createSecureRequest(
        payload: String,
        sessionKey: SecretKey,
        requestMetadata: RequestMetadata
    ): SecureRequest? {
        // Create AAD from request metadata for additional authentication
        val aad = "${requestMetadata.method}|${requestMetadata.uri}|${requestMetadata.timestamp}"
        
        // Encrypt the payload
        val encryptedPayload = encryptPayload(payload, sessionKey, aad) ?: return null
        
        return SecureRequest(
            encryptedPayload = encryptedPayload,
            metadata = requestMetadata
        )
    }
    
    /**
     * Validates and extracts payload from a secure response.
     * 
     * @param secureResponse The secure response envelope
     * @param sessionKey Session key for decryption
     * @param expectedMetadata Expected metadata for validation
     * @return Decrypted response payload, or null on failure
     */
    fun extractSecureResponse(
        secureResponse: SecureResponse,
        sessionKey: SecretKey,
        expectedMetadata: ResponseMetadata? = null
    ): String? {
        // Validate metadata if provided
        expectedMetadata?.let {
            if (secureResponse.metadata.statusCode != it.statusCode) {
                Log.e(TAG, "Response metadata mismatch")
                return null
            }
        }
        
        // Decrypt the payload
        return decryptPayload(secureResponse.encryptedPayload, sessionKey)
    }
    
    /**
     * Data class representing an encrypted payload.
     */
    data class EncryptedPayload(
        val version: Int,
        val iv: String,           // Base64 encoded initialization vector
        val ciphertext: String,    // Base64 encoded ciphertext (includes auth tag)
        val algorithm: String,
        val aad: String? = null    // Associated authenticated data (if any)
    )
    
    /**
     * Request metadata for additional security context.
     */
    data class RequestMetadata(
        val method: String,
        val uri: String,
        val timestamp: Long,
        val sessionId: String
    )
    
    /**
     * Response metadata for validation.
     */
    data class ResponseMetadata(
        val statusCode: Int,
        val timestamp: Long,
        val sessionId: String
    )
    
    /**
     * Secure request envelope containing encrypted payload and metadata.
     */
    data class SecureRequest(
        val encryptedPayload: EncryptedPayload,
        val metadata: RequestMetadata
    )
    
    /**
     * Secure response envelope containing encrypted payload and metadata.
     */
    data class SecureResponse(
        val encryptedPayload: EncryptedPayload,
        val metadata: ResponseMetadata
    )
}